### PR TITLE
Add a functional test for file uploads

### PR DESCRIPTION
**Pull request by di**
This PR adds a functional test which performs a `POST` with a file upload.

### CODE REVIEW

File: tests/common/db/macaroons.py [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+# limitations under the License.
+
+import factory
+
+from warehouse.macaroons.models import Macaroon
+
+from .base import WarehouseFactory
+
+
+class MacaroonFactory(WarehouseFactory):
```
Comment (miketheman) [inline tests/common/db/macaroons.py:20]: ✨ Yay! A new 🏭 Factory!

File: tests/conftest.py [hunk 2]
```diff
@@ -333,6 +333,7 @@ def get_app_config(database, nondefaults=None):
         "sessions.url": "redis://localhost:0/",
         "statuspage.url": "https://2p66nmmycsj3.statuspage.io",
         "warehouse.xmlrpc.cache.url": "redis://localhost:0/",
+        "archive_files.path": "/tmp",
```
Comment (miketheman) [inline tests/conftest.py:336]: nit: move this up, right beneath `"archive_files.backend"`

File: tests/functional/_fixtures/sampleproject-3.0.0.tar.gz
Comment (miketheman) [inline tests/functional/_fixtures/sampleproject-3.0.0.tar.gz:1]: It'd be nice for us to have something somewhere on what's in this, and how to make a new one

File: tests/functional/forklift/test_legacy.py [hunk 3]
```diff
@@ -48,3 +56,61 @@ def test_remove_doc_upload(webtest):
...
+
+
+def test_file_upload(webtest):
+    user = UserFactory.create(
+        password=(  # 'password'
+            "$argon2id$v=19$m=1024,t=6,p=6$EiLE2Nsbo9S6N+acs/beGw$ccyZDCZstr1/+Y/1s3BVZ"
+            "HOJaqfBroT0JCieHug281c"
+        )
+    )
+    EmailFactory.create(user=user, verified=True)
```
Comment (miketheman) [inline tests/functional/forklift/test_legacy.py:68]: We have a UserFactory trait now, would that accomplish your needs?
```suggestion
    user = UserFactory.create(with_verified_primary_email=True)
```

What purpose does setting a hashed password value here? Could we supply a string instead, or does some other mechanic prevent this?

**Review by miketheman — COMMENTED** [review]
Yay, functional testing at play!

Left a couple of inline notes.

The [test now appears on the "slow" list](https://github.com/pypi/warehouse/actions/runs/10621734746/job/29444571167#step:5:215) - this test takes ~1.2 seconds - I wonder what the slow part is?

File: tests/functional/forklift/test_legacy.py (hunk 3 above)
Comment (di) [reply tests/functional/forklift/test_legacy.py:68]: I was having an issue where the factory was not actually making this a valid 'enabled' password, and the user account was not usable as a result. Maybe this should be a default value? I think we can definitely use the trait though.

File: tests/functional/_fixtures/sampleproject-3.0.0.tar.gz
Comment (di) [reply tests/functional/_fixtures/sampleproject-3.0.0.tar.gz:1]: This is just from https://pypi.org/project/sampleproject/3.0.0/#files. Where would you expect to find those details? Would inline with the test be OK?

File: tests/functional/forklift/test_legacy.py (hunk 3 above)
Comment (di) [reply tests/functional/forklift/test_legacy.py:68]: Yeah this is why: https://github.com/pypi/warehouse/blob/ccff64920db7965078cf1fdb50f028e640328887/tests/common/db/accounts.py#L39

Causes this to fail: https://github.com/pypi/warehouse/blob/ccff64920db7965078cf1fdb50f028e640328887/warehouse/accounts/services.py#L342

File: tests/functional/_fixtures/sampleproject-3.0.0.tar.gz
Comment (miketheman) [reply tests/functional/_fixtures/sampleproject-3.0.0.tar.gz:1]: Ah excellent - that's cool. A link in the test case docstring would be where I'd look for it, or in a README file in the _fixtures subdir next to this file

Comment (di) [reply tests/functional/_fixtures/sampleproject-3.0.0.tar.gz:1]: I added a `README.md` to the fixtures directory.

File: tests/functional/forklift/test_legacy.py (hunk 3 above)
Comment (miketheman) [reply tests/functional/forklift/test_legacy.py]: Yeah, the default password is `"!"` which is unix-y for disabled. We could probably change our default and see what happens!

**Comment by di** [comment]
> The test now appears on the "slow" list - this test takes ~1.2 seconds - I wonder what the slow part is?

Probably the fact that this is invoking the entire upload view, which is pretty heavy, plus doing a file upload.

**Review by miketheman — APPROVED** [review]
Nice stuff! Feel free to explore the password issue now or later, but this is good to go

**Review by miketheman — APPROVED** [review]

**Pull request merged**